### ISSUE TITLE
`cpu(x)` errors for `x isa CuArray{<:CartesianIndex}`

### ISSUE THREAD

**Issue by Sleort** [FluxML/Flux.jl#2116]
### Package Version

Flux v0.13.7

### Julia Version

1.8.2

### OS / Environment

Ubuntu 20.04

### Describe the bug

I can copy an `Array` of `CartesianIndex`s to the GPU using Flux's convenience function `gpu()`, but not back to the CPU again using `cpu()`. The `Array` constructor works, though. See the MWE below.

### Steps to Reproduce

```
julia> using Flux

julia> x_cpu = [CartesianIndex(1)]
1-element Vector{CartesianIndex{1}}:
 CartesianIndex(1,)

julia> x_gpu = gpu(x_cpu)
1-element CUDA.CuArray{CartesianIndex{1}, 1, CUDA.Mem.DeviceBuffer}:
 CartesianIndex(1,)

julia> Array(x_gpu)    #Works
1-element Vector{CartesianIndex{1}}:
 CartesianIndex(1,)

julia> cpu(x_gpu)    #Does not work
ERROR: GPU broadcast resulted in non-concrete element type Any.
This probably means that the function you are broadcasting contains an error or type instability.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] copy
    @ ~/.julia/packages/GPUArrays/fqD8z/src/host/broadcast.jl:34 [inlined]
  [3] materialize(bc::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{1}, Nothing, Functors.var"#16#17"{typeof(Functors.isleaf), typeof(Functors._default_walk), IdDict{Any, Any}, Functors.NoKeyword, Flux.var"#160#161"}, Tuple{CUDA.CuArray{CartesianIndex{1}, 1, CUDA.Mem.DeviceBuffer}}})
    @ Base.Broadcast ./broadcast.jl:860
  [4] map(::Function, ::CUDA.CuArray{CartesianIndex{1}, 1, CUDA.Mem.DeviceBuffer})
    @ GPUArrays ~/.julia/packages/GPUArrays/fqD8z/src/host/broadcast.jl:82
  [5] _default_walk
    @ ~/.julia/packages/Functors/V2McK/src/functor.jl:39 [inlined]
  [6] fmap(f::Flux.var"#160#161", x::CUDA.CuArray{CartesianIndex{1}, 1, CUDA.Mem.DeviceBuffer}; exclude::typeof(Functors.isleaf), walk::typeof(Functors._default_walk), cache::IdDict{Any, Any}, prune::Functors.NoKeyword)
    @ Functors ~/.julia/packages/Functors/V2McK/src/functor.jl:46
  [7] fmap
    @ ~/.julia/packages/Functors/V2McK/src/functor.jl:44 [inlined]
  [8] cpu(x::CUDA.CuArray{CartesianIndex{1}, 1, CUDA.Mem.DeviceBuffer})
    @ Flux ~/.julia/packages/Flux/nJ0IB/src/functor.jl:154
  [9] top-level scope
    @ REPL[4]:1
 [10] top-level scope
    @ ~/.julia/packages/CUDA/DfvRa/src/initialization.jl:52
```

### Expected Results

Well, I expected to get `x_cpu` back; `x_cpu == cpu(gpu(x_cpu))`.

### Observed Results

`cpu(gpu(x_cpu))` errors for `x_cpu isa Array{<:CartesianIndex}` .

**Comment by DhairyaLGandhi** [comment]
One curious thing is the call to `map(f, ::CuArray{CartesianIndex})`. It's almost as if we are calling `map(Array, x_gpu)` which is incorrect in this case.

Would you be able to post the result of `@code_warntype cpu(::CuArray{CartesianIndex})` and also try with `@functor CartesianIndex`.

**Comment by CarloLucibello** [comment]
The issue is whenever `CuArray{T}` where T is non-numerical immutable type. The issue will be fixed in #2117
